### ISSUE TITLE
AoE+Harvest drops grass blocks

### ISSUE THREAD

**Issue by EoD** [Mithion/ArsMagica2#1263]
Using AM2 80ba470ed23de9edec8415672d7758d4c182d767

Using AoE+Harvest drops grass blocks.
![2015-09-20_18 41 10](https://cloud.githubusercontent.com/assets/293499/9981789/54fdffa4-5fc7-11e5-8214-12063114723c.png)

**Comment by Kolatra** [comment]
It looks like it's dropping dirt.

Do you have a better screenshot?

**Comment by EoD** [comment]
Not really, but it is really simple to reproduce.

When you use AoE+Harvest, all grass blocks with entities on it (flowers, tall grass, etc) in the radius of the spell will drop as dirt blocks.

**Comment by Kolatra** [comment]
I was not able to reproduce this.

**Comment by jjtParadox** [comment]
I was kind of able to reproduce this, but the issue I found is not the same as the issue labeled above.

![2015-09-21_10 22 19](https://cloud.githubusercontent.com/assets/1388166/9996079/0863089a-604b-11e5-8f52-31d8d2772f5c.png)

If standing in the hole in the grass blocks on the left, casting AoE+Harvest will break all the grass blocks, as if by a Dig spell.

![2015-09-21_10 22 32](https://cloud.githubusercontent.com/assets/1388166/9996113/2d2d3948-604b-11e5-9a31-d0fb28fe67d5.png)
![2015-09-21_10 22 50](https://cloud.githubusercontent.com/assets/1388166/9996117/3074ccb0-604b-11e5-8b97-38f11f1b6fa2.png)

But the cobblestone on the right is unaffected by the spell.

![2015-09-21_10 22 55](https://cloud.githubusercontent.com/assets/1388166/9996132/44a2055e-604b-11e5-8577-31b867d3bd82.png)

If the grass is replaced with dirt, the dirt is unaffected.

![2015-09-21_10 23 37](https://cloud.githubusercontent.com/assets/1388166/9996141/59b086a0-604b-11e5-85a8-8010db38c3b9.png)

What is happening here is that grass counts as a "growable" block, so the harvest spell is targeting it and breaking it as if it were a crop.

**Comment by EoD** [comment]
Yeah, precisely as you (and me?) wrote: Only grass blocks are being dropped.

**Comment by jjtParadox** [comment]
I think there was some confusion over what you meant by "grass blocks are being dropped". Worded that way, it sounds as if silk touch is being used on the grass.

I think a clearer description is that the harvest spell is breaking grass blocks, which seems wrong as grass blocks are not crops, or "harvest-able" in the sense that wheat and pumpkins are.

It isn't an issue with a spell using silk touch on a block, but that a spell is affecting something that it probably shouldn't be affecting.

**Comment by EoD** [comment]
Thanks, I updated the initial description and the title to clarify the bug report.

**Comment by Mithion** [comment]
Perhaps a configurable harvest blacklist is in order?
I could see certain server admins not wanting harvest to work on certain plants.  Then we could just add grass blocks as a default in the config.

**Comment by jjtParadox** [comment]
That is a possibility, but there may be room for a cleaner solution that doesn't require server admins manually blacklisting whatever varieties of grass that are added by other mods. Possibly checking other conditions instead of just `instanceof IGrowable`?

Looking at some of the base MC code, I may have already found a bug in that Melons and Pumpkins don't implement IGrowable, so casting harvest on them may just break the stem, not the pumpkin or melon itself.

**Comment by Mithion** [comment]
If there's a cleaner solution I'm all for it, but I could still see thematically or for various balance reasons admins wanting to make it that certain blocks can't be broken - so maybe not a priority but a config would still fall under a "nice to have".

**Comment by EoD** [comment]
A simple check would be to check for ``Material == Material.plants``. This works at least for vanilla MC.

**Comment by Mithion** [comment]
So something like this at the beginning of Harvest:

if ( !(block instanceof IGrowable) || block.isNormalCube() ){
    return;
}

**Comment by EoD** [comment]
Is the Harvest spell supposed to break saplings, too?

**Comment by jjtParadox** [comment]
Issue with `block.isNormalCube()` is that melons count as normal cubes.

I'm starting to really wonder at how other mods (Thaumcraft, I'm looking at you) figure out what to break with a harvester and what not to break.

**Comment by Mithion** [comment]
Thaumcraft has many other mods make use of its API as well though.  

I say we come up with a general set of rules that handles *most* cases, like my example above, then allow registering custom cases via both a whitelist to add things that aren't covered, and a blacklist to cover things that the modder doesn't want the generic case to catch.

To handle things like melons, we add to said whitelist by default.

**Comment by Mithion** [comment]
To clarify, the whitelist/blacklist would be accessible via the AM2 API.

**Comment by EoD** [comment]
``BlockMelon`` does not implement ``IGrowable``, hence it does not break with the Harvest component anyway.

**Comment by EoD** [comment]
Even though Vexatos voted against it, another way would be to check for ``IGrowable`` together with a ``material == Material.plants`` check. This then includes saplings, flowers, crops and many other similar things.

**Comment by jjtParadox** [comment]
Whoops I completely forgot about this thread :worried: 

Looking into some code from tech mods that have harvesters (namely MineFactoryReloaded), @Mithion's proposal of an API with a simple whitelist/blacklist may be what's necessary for the harvest component to work properly.
